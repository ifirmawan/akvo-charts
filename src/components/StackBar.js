import React from 'react';
import { useECharts } from '../hooks';
import styles from '../styles.module.css';
import { Tooltip } from '../utils/basicChartStyle';

const getOptions = ({ dimensions, stackMapping, horizontal = true }) => {
  // Reverse the stackMapping to get a dimension to stack group map
  const dimensionToStackMap = {};
  Object.keys(stackMapping).forEach((stackGroup) => {
    stackMapping[stackGroup].forEach((dim) => {
      dimensionToStackMap[dim] = stackGroup;
    });
  });

  // Create series based on the reversed stack mapping
  const series = dimensions.slice(1).map((dim) => ({
    name: dim,
    type: 'bar',
    stack: dimensionToStackMap[dim] || 'defaultStack',
    encode: {
      x: horizontal ? dim : 'category',
      y: horizontal ? 'category' : dim
    }
  }));

  return {
    tooltip: {
      ...Tooltip,
      trigger: 'axis'
    },
    series
  };
};

const StackBar = ({ config, data, stackMapping = {}, horizontal = true }) => {
  const chartRef = useECharts({
    config: { ...config, horizontal },
    data,
    getOptions: ({ dimensions }) =>
      getOptions({ dimensions, stackMapping, horizontal })
  });

  return (
    <div
      ref={chartRef}
      role="figure"
      className={styles.container}
    />
  );
};

export default StackBar;
