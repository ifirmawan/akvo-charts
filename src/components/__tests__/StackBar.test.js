import React from 'react';
import { render, screen } from '@testing-library/react';
import StackBar from '../StackBar';
import renderer from 'react-test-renderer';

describe('StackBar chart', () => {
  test('renders StackBar component with 2d array data format', () => {
    const data = [
      ['product', '2015', '2016', '2017'],
      ['Matcha Latte', 43.3, 85.8, 93.7],
      ['Milk Tea', 83.1, 73.4, 55.1],
      ['Cheese Cocoa', 86.4, 65.2, 82.5],
      ['Walnut Brownie', 72.4, 53.9, 39.1]
    ];

    const config = {
      title: 'StackBar Chart Example',
      xAxisLabel: 'Categories',
      yAxisLabel: 'Values'
    };

    render(
      <StackBar
        config={config}
        data={data}
      />
    );

    const chartContainer = screen.getByRole('figure');
    expect(chartContainer).toBeInTheDocument();
  });

  test('renders StackBar component with row based key-value format (object array)', () => {
    const data = [
      { product: 'Matcha Latte', count: 823, score: 95.8 },
      { product: 'Milk Tea', count: 235, score: 81.4 },
      { product: 'Cheese Cocoa', count: 1042, score: 91.2 },
      { product: 'Walnut Brownie', count: 988, score: 76.9 }
    ];

    const config = {
      title: 'StackBar Chart Example',
      xAxisLabel: 'Categories',
      yAxisLabel: 'Values'
    };

    render(
      <StackBar
        config={config}
        data={data}
      />
    );

    const chartContainer = screen.getByRole('figure');
    expect(chartContainer).toBeInTheDocument();
  });

  test('renders StackBar component with column based key-value format', () => {
    const data = {
      product: ['Matcha Latte', 'Milk Tea', 'Cheese Cocoa', 'Walnut Brownie'],
      count: [823, 235, 1042, 988],
      score: [95.8, 81.4, 91.2, 76.9]
    };

    const config = {
      title: 'StackBar Chart Example',
      xAxisLabel: 'Categories',
      yAxisLabel: 'Values'
    };

    const stackMapping = {
      group1: [2015, 2016],
      group2: [2017, 2018]
    };

    render(
      <StackBar
        config={config}
        data={data}
        stackMapping={stackMapping}
        horizontal={false}
      />
    );

    const chartContainer = screen.getByRole('figure');
    expect(chartContainer).toBeInTheDocument();
  });

  test('matches StackBar snapshot', () => {
    const data = [
      ['product', '2015', '2016', '2017'],
      ['Matcha Latte', 43.3, 85.8, 93.7],
      ['Milk Tea', 83.1, 73.4, 55.1],
      ['Cheese Cocoa', 86.4, 65.2, 82.5],
      ['Walnut Brownie', 72.4, 53.9, 39.1]
    ];

    const config = {
      title: 'StackBar Chart Example',
      xAxisLabel: 'Categories',
      yAxisLabel: 'Values'
    };

    const tree = renderer
      .create(
        <StackBar
          config={config}
          data={data}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
