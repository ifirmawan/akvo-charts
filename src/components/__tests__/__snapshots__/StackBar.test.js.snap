// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackBar chart matches StackBar snapshot 1`] = `
<div
  className="container"
  role="figure"
/>
`;
